org.civicrm.hremergency/tests/phpunit/api/v3/Contact/DeleteemergencycontactTest.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+use Civi\Test\HeadlessInterface;
+use Civi\Test\TransactionalInterface;
+use Tests\CiviHR\HREmergency\Fabricator\EmergencyContactFabricator;
+
+/**
+ * @group headless
+ */
+class api_v3_Contact_DeleteEmergencyContactTest extends \PHPUnit_Framework_TestCase
+  implements HeadlessInterface, TransactionalInterface {
```

Comment:
So here we have a common problem whenever we need to test the API: not being able to inject things. The problem is that we're only testing the API, not the Service that is using. Usually, in cases like this, I would recommend to focus the tests on the service and, for the API, check only what it is specific to it (like the permission validations). However, since you've exposed the service in the CiviCRM container, I think we can use that in our favor. I think that for the API tests you can replace the actual service with a mock (using `Civi::container->set()`) which avoid the duplication between the service and the API tests

Reply:
I considered this. I've done this kind of thing before, eventually _all_ tests needed some setup where the container was modified which I would like to avoid. For example

```
public function testCreate() {
  $mockService = $this->prophesize(EmergencyContactService::class);
  $mockService->find(1)->willReturn(['id' => 1, 'name' => 'foo']);
  Civi::container()->set('emergency_contact.service', $mockService->reveal());
  // start the test
}
```
It would be a lot nicer for testing services with other dependencies to inject those mocked dependencies directly instead of altering the container every time.

That said, I don't think this is avoidable when we're doing function calls like `civicrm_api3` or calling anything that does static calls, so I can change this test to use `$container->set()`